doc/cookbook/guard_authentication.rst

```diff
@@ -0,0 +1,150 @@
+How to Create a Custom Authentication System with Guard
+=======================================================
+
+Whether you need to build a traditional login form, an API token authentication system or you need to integrate with some proprietary single-sign-on system, the Guard component can make it easy... and fun!
+
+In this example, you'll build an API token authentication system and learn how to work with Guard.
+
+Step 1) Create the Authenticator Class
+--------------------------------------
+
+Suppose you have an API where your clients will send an X-AUTH-TOKEN header on each request. This token is composed of the username followed by a password, separated by a colon. Your job is to read this, find the associated user (if any) and check the password.
+
+To create a custom authentication system, just create a class and make it implement GuardAuthenticatorInterface. Or, extend the simpler AbstractGuardAuthenticator. This requires you to implement six methods:
+
+.. code-block:: php
+
+    <?php
+
+    namespace App\Security;
+
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\JsonResponse;
+    use Symfony\Component\Security\Core\User\UserInterface;
+    use Symfony\Component\Security\Core\User\UserProviderInterface;
+    use Symfony\Component\Security\Guard\AbstractGuardAuthenticator;
+    use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+    use Symfony\Component\Security\Core\Exception\AuthenticationException;
+
+    class TokenAuthenticator extends AbstractGuardAuthenticator
+    {
+        public function getCredentials(Request $request)
+        {
+            if (!$token = $request->headers->get('X-AUTH-TOKEN')) {
+                return;
+            }
+
+            list($username, $secret) = explode(':', $token, 2);
+
+            return array(
+                'username' => $username,
+                'secret' => $secret,
+            );
+        }
+
+        public function getUser($credentials, UserProviderInterface $userProvider)
+        {
+            return $userProvider->loadUserByUsername($credentials['username']);
+        }
+
+        public function checkCredentials($credentials, UserInterface $user)
+        {
+            // check credentials - e.g. make sure the password is valid
+            // no credential check is needed in this case
+
+            // return true to cause authentication success
+            return $user->getPassword() === $credentials['secret'];
```

Comment:
Is the `secret` (the password) supposed to be encoded when the user sends it? If it's plain-text, then we would need to encode that secret to compare it with the encoded password, right?